src/libs/actions/Report.ts

```diff
@@ -1086,50 +1093,82 @@ function openReport(
-    if (transactionID && !parentReportActionID) {
-        const transaction = allTransactions?.[transactionID];
-
-        if (transaction) {
-            const selfDMReportID = findSelfDMReportID();
-
-            if (selfDMReportID) {
-                const generatedReportActionID = rand64();
-                const optimisticParentAction = buildOptimisticIOUReportAction({
-                    type: CONST.IOU.REPORT_ACTION_TYPE.CREATE,
-                    amount: Math.abs(transaction.amount),
-                    currency: transaction.currency,
-                    comment: transaction.comment?.comment ?? '',
-                    participants: [{accountID: currentUserAccountID, login: currentUserEmail ?? ''}],
-                    transactionID,
-                    isOwnPolicyExpenseChat: true,
-                });
+    // This is a legacy transaction that doesn't have either a transaction thread or a money request preview
+    if (transaction && !parentReportActionID) {
+        const isUnreportedTransaction = transaction?.reportID === CONST.REPORT.UNREPORTED_REPORT_ID;
+        const parentReportID = isUnreportedTransaction ? findSelfDMReportID() : transaction?.reportID;
+        const iouReportActionID = rand64();
+
+        const optimisticIOUAction = buildOptimisticIOUReportAction({
+            type: CONST.IOU.REPORT_ACTION_TYPE.CREATE,
+            amount: Math.abs(transaction.amount),
+            currency: transaction.currency,
+            comment: transaction.comment?.comment ?? '',
+            participants: [{accountID: currentUserAccountID, login: currentUserEmail ?? ''}],
+            transactionID: transaction.transactionID,
+            isOwnPolicyExpenseChat: true,
+        });
 
-                optimisticData.push({
-                    onyxMethod: Onyx.METHOD.MERGE,
-                    key: `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
-                    value: {
-                        parentReportID: selfDMReportID,
-                        parentReportActionID: generatedReportActionID,
-                    },
-                });
+        // We have a case where the transaction data is only available from the snapshot
+        // So we need to add the transaction data to Onyx so it's available when opening the report
+        optimisticData.push({
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.TRANSACTION}${transaction.transactionID}`,
+            value: transaction,
+        });
 
-                optimisticData.push({
-                    onyxMethod: Onyx.METHOD.SET,
-                    key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${selfDMReportID}${generatedReportActionID}`,
-                    value: {
-                        ...optimisticParentAction,
-                        reportActionID: generatedReportActionID,
-                        childReportID: reportID,
-                    },
-                });
+        // Add violations if they exist
```

Comment:
So if the violations are passed, it means that is only in the snapshot? That is why we need to add it to normal onyx? 

Could you please update the comment to make that clear? 🙏